core/controllers/profile.py

```diff
@@ -258,26 +258,275 @@ def put(self) -> None:
+                        'schema': {
+                            'type': 'bool'
+                        }
+                    }]
+                }
+            }
+        }
+    }
 
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
 
     @acl_decorators.can_manage_own_account
     def get(self) -> None:
         """Handles GET requests."""
         assert self.user_id is not None
-        user_settings = user_services.get_user_settings(self.user_id)
         user_email_preferences = user_services.get_email_preferences(
             self.user_id)
+        self.values.update({
+            'can_receive_email_updates': (
+                user_email_preferences.can_receive_email_updates),
+            'can_receive_editor_role_email': (
+                user_email_preferences.can_receive_editor_role_email),
+            'can_receive_feedback_message_email': (
+                user_email_preferences.can_receive_feedback_message_email),
+            'can_receive_subscription_email': (
+                user_email_preferences.can_receive_subscription_email)
+        })
+        self.render_json(self.values)
+
+    @acl_decorators.can_manage_own_account
+    def put(self) -> None:
+        """Handles PUT requests."""
+        assert self.normalized_payload is not None
+        assert self.user_id is not None
+        data = self.normalized_payload['data']
+        bulk_email_signup_message_should_be_shown = (
+            user_services.update_email_preferences(
+                self.user_id, data['can_receive_email_updates'],
+                data['can_receive_editor_role_email'],
+                data['can_receive_feedback_message_email'],
+                data['can_receive_subscription_email']
+            )
+        )
+
+        self.render_json({
+            'bulk_email_signup_message_should_be_shown': (
+                bulk_email_signup_message_should_be_shown)
+        })
+
+
+class ProfilePictureDataUrlHandler(
+    base.BaseHandler[Dict[str, str], Dict[str, str]]
+):
+    """Handles requests for the profile picture data url."""
+
+    URL_PATH_ARGS_SCHEMAS: Dict[str, str] = {}
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {},
+        'PUT': {
+            'data': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            }
+        }
+    }
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    @acl_decorators.can_manage_own_account
+    def get(self) -> None:
+        """Handles GET requests."""
+        assert self.user_id is not None
+        user_settings = user_services.get_user_settings(self.user_id)
+
+        self.values.update({
+            'profile_picture_data_url': user_settings.profile_picture_data_url
+        })
+        self.render_json(self.values)
+
+    @acl_decorators.can_manage_own_account
+    def put(self) -> None:
+        """Handles PUT requests."""
+        assert self.user_id is not None
+        assert self.normalized_payload is not None
```

Comment:
Is this needed?

Reply:
I think so. We are inheriting `normalized_payload` from `BaseHandler` and [there](https://github.com/oppia/oppia/blob/ac68b404fef555b6982613effc483261365ea147/core/controllers/base.py#L220) it is set to `Optional`. But here we are not expecting `None` so, to cancel out that probability I think we need this.